File: 20_notvalid.cpp
class Solution {
public:
    bool isValid(string s) {
        
        if(!s.size()) return true;
        stack<int> brackets;
        
        for(int i=0; i<s.size(); i++){
            
            char c = s[i];
            if(c=='(' or c=='{' or c=='[') brackets.push(c);
            else if(brackets.empty() and (c==')' or c=='}' or c==']')) brackets.push(c);
            else {
                if(c==')' and brackets.top()=='(') brackets.pop();
                else if(c=='}' and brackets.top()=='{') brackets.pop();
                else if(c==']' and brackets.top()=='[') brackets.pop();
                else brackets.push(c);
            }
        }
        return brackets.empty()? true: false;
    }
};

File: 20_notvalid.py
#append to a list and check if there is a pair
class Solution:
		def isValid(self, s: str) -> bool:
			if not s:
				return True
			if len(s)==1:
				return False
			left = "{[("
			pair = "{}()[]"
			x = []
			for i in s:
				if i in left:
					x.append(i)
				else:
					if not x or x.pop()+i not in pair:
						return False      
			return True if not x else False